lib/extensions.js

```diff
@@ -776,4 +778,34 @@ module.exports = function(realmConstructor, context) {
         },
         configurable: true
     });
+
+    Object.defineProperty(realmConstructor, "JsonSerializationReplacer", {
+        get: function () {
+            const seen = [];
+
+            return function(_, value) {
+                if (typeof value !== "object") {
+                    return value;
+                }
+
+                const parent = this;
+                if (seen.length > 0) {
+                    const pos = seen.indexOf(parent);
+                    if (pos !== -1) {
+                        seen.splice(pos + 1);
+                    } else {
+                        seen.push(parent);
+                    }
+
+                    if (seen.includes(value)) {
+                        return value.$realmId ? { $ref: value.$realmId } : "[Circular reference]";
+                    }
+                } else {
+                    seen.push(value);
+                }
```

Comment:
This section of the code definitely needs comments - I'm having a very hard time wrapping my head around it and confirming it's behaving as expected, especially considering it's modifying some external array.

Reply:
Yes definitely, it's on the todo - maybe the draft/WIP was opened a bit early, just to show the overall approach.

Reply:
The PR description was longer than the PR itself, so of course I completely skipped it 😅

Reply:
Added comments.